src/networking/nc/mod.rs

```diff
@@ -0,0 +1,958 @@
+use clap::{Arg, ArgMatches};
+use mio::{Events, Event, Poll, Ready, PollOpt, Token};
+use libc::{AF_UNSPEC, AF_INET, AF_INET6, AF_UNIX};
+use std::io;
+use std::net::{SocketAddr};
+use mio::unix::EventedFd;
+use std::io::{Read,Write, ErrorKind};
+use mio::unix::UnixReady;
+use std::thread::sleep;
+use std::fmt::Debug;
+use std::os::unix::io::AsRawFd;
+use std::os::unix::io::RawFd;
+use std::time::Duration;
+use std::net::{ToSocketAddrs};
+use socket2::{Socket, Domain};
+use super::{UtilSetup, ArgsIter};
+use super::{MesaError};
+
+
+pub(crate) const NAME: &str = "nc";
+pub(crate) const DESCRIPTION: &str = "arbitrary TCP and UDP connections and listens";
+
+const BUFSIZE: usize = 16384;
+const PRINT_DEBUG_INFO: bool = false;
+
+#[derive(Debug)]
+struct NcOptions {
+    dflag: bool,
+    iflag: bool,
+    interval: Option<Duration>,
+    lflag: bool,
+    host: String,
+    portlist: Vec<u16>,
+    unixflag: bool,
+    uflag: bool,
+    family: i32,
+    kflag: bool,
+    s_addr: Option<String>,
+    pflag: bool,
+    vflag: bool,
+    zflag: bool,
+    timeout: Option<Duration>,
+    unix_dg_tmp_socket: String,
+}
+
+fn mesaerr_result<T>(err_msg: &str) -> Result<T, MesaError> {
+    Err(MesaError::from(
+        failure::err_msg(format!("{}", err_msg)).compat()
+        ))
+}
+
+fn build_ports(ports: &str) -> Result<Vec<u16>, MesaError>{
+    // TODO: suport XX-XX
+    let port_list = match ports.parse::<u16>() {
+        Ok(port) => port,
+        Err(_) => {
+            return mesaerr_result(&format!("invalid port[s] {}", ports));
+        }
+    };
+
+    Ok(vec!(port_list))
+}
+
+fn warn(msg: &str) {
+    eprint!("{}", msg);
+}
+
+fn debug_info(msg: &str) {
+    if PRINT_DEBUG_INFO {
+        eprint!("{}", msg);
+    }
+}
+
+impl NcOptions {
+    pub fn parse(matches: ArgMatches, msg: &str) -> Result<NcOptions, MesaError> {
+        let mut portlist = vec!();
+        let lflag = matches.is_present("l");
+        let mut host = String::from("127.0.0.1");
+        let uport:String;
+        let mut interval = None;
+        let mut timeout = None;
+        let s_addr = match matches.value_of("s") {
+            Some(addr) => Some(String::from(addr)),
+            None => None
+        };
+        let uflag = matches.is_present("u");
+        let pflag = matches.is_present("p");
+
+        let zflag = matches.is_present("z");
+        let kflag = matches.is_present("k");
+
+        /* Cruft to make sure options are clean, and used properly. */
+        let positionals: Vec<&str> = if matches.is_present("positionals") {
+            matches.values_of("positionals").unwrap().collect()
+        } else {
+            vec!()
+        };
+
+        let family = if matches.is_present("U") {
+            AF_UNIX
+        } else if matches.is_present("6") {
+            AF_INET6
+        } else if matches.is_present("4") {
+            AF_INET
+        } else {
+            AF_UNSPEC
+        };
+
+        if positionals.len() == 1 {
```

Comment:
```rust
// NOTE: i'm pretty sure all of these can be &str rather than String if you
//       restructure this function a bit and give NcOptions a lifetime
let first_pos = String::from(positionals.next().unwrap());
let (host, uport) = match positionals.next() {
    Some(uport) => (first_pos, uport),
    None => if family == AF_UNIX {
        (first_pos, String::new())
    } else {
        // there may be a way to configure clap so it can handle this,
        // but it's probably not simple
        if !lflag {
            return mesaerr_result(msg);
        }
        (String::from("127.0.0.1"), first_pos)
    }
};
```